dashboard/modules/job/job_head.py

```diff
@@ -57,32 +83,81 @@ async def get_package(self,
     async def upload_package(self, req: aiohttp.web.Request):
         package_uri = req.query["package_uri"]
         logger.info(f"Uploading package {package_uri} to the GCS.")
-        upload_package_to_gcs(package_uri, await req.read())
+        try:
+            upload_package_to_gcs(package_uri, await req.read())
+        except Exception:
+            return aiohttp.web.Response(
+                reason=traceback.format_exc().encode("utf-8"),
+                status=aiohttp.web.HTTPInternalServerError.status_code)
 
-        return aiohttp.web.Response()
+        return aiohttp.web.Response(
+            text={"success": True},
+            content_type="application/json",
+            status=aiohttp.web.HTTPOk.status_code,
+        )
 
     @routes.post(JOBS_API_ROUTE_SUBMIT)
     @_ensure_ray_initialized
     async def submit(self, req: aiohttp.web.Request) -> aiohttp.web.Response:
-        # TODO: (jiaodong) Validate if job request is valid without using
-        # pydantic.
-        submit_request = JobSubmitRequest(**(await req.json()))
-        job_id = self._job_manager.submit_job(
-            entrypoint=submit_request.entrypoint,
-            runtime_env=submit_request.runtime_env,
-            metadata=submit_request.metadata)
-
-        resp = JobSubmitResponse(job_id=job_id)
+        result = await self._parse_and_validate_request(req, JobSubmitRequest)
+        # Request parsing failed, returned with Response object.
+        if isinstance(result, aiohttp.web.Response):
+            return result
+        else:
+            submit_request = result
+
+        try:
+            job_id = self._job_manager.submit_job(
+                entrypoint=submit_request.entrypoint,
+                runtime_env=submit_request.runtime_env,
+                metadata=submit_request.metadata)
+
+            resp = JobSubmitResponse(job_id=job_id)
+        except Exception:
+            return aiohttp.web.Response(
+                reason=traceback.format_exc().encode("utf-8"),
+                status=aiohttp.web.HTTPInternalServerError.status_code)
+
+        return aiohttp.web.Response(
+            text=json.dumps(dataclasses.asdict(resp)),
+            content_type="application/json",
+            status=aiohttp.web.HTTPOk.status_code,
+        )
+
+    @routes.delete(JOBS_API_ROUTE_STOP)
```

Comment:
I don't think this is really a `DELETE` because we don't clean up the job record. I don't know exactly what the idiomatic method to choose here would be though. @tchordia any thoughts?

Reply:
our "stop" api is fire and forget so hard to guarantee to be idempotent. `POST` makes sense too.